### PR TITLE
feat: add `_cast_list_f16` function for casting list of floats to f16

### PR DESCRIPTION
**Pull request by chebbyChefNEQ**
Add UDF for casting vector columns

### CODE REVIEW

File: rust/lance/src/dataset/scanner.rs [hunk 1]
```diff
@@ -2875,6 +2876,60 @@ mod test {
...
+            .await
+            .unwrap();
+
+        let dataset = Dataset::open(test_uri).await.unwrap();
+        assert_eq!(32, dataset.scan().count_rows().await.unwrap());
+
+        let mut scanner = dataset.scan();
+
+        let scan_res = scanner
+            .project_with_transform(&[("f16", "cast_vector_f16(vec)")])
```
Comment (eddyxu) [inline rust/lance/src/dataset/scanner.rs:2895]: we can not use the same function but different kernels with different source type?

Comment (chebbyChefNEQ) [reply rust/lance/src/dataset/scanner.rs:2895]: this works for fsl of f32 and f16. Probably good enough for now? It's pretty easy to tweak the UDF to support more

Comment (eddyxu) [reply rust/lance/src/dataset/scanner.rs:2895]: ohi i mean can we do `cast_f16` for both f32 or `list[f32]` or `fsl[f32]`? Dont have the "vector" in udf name

Comment (chebbyChefNEQ) [reply rust/lance/src/dataset/scanner.rs:2895]: oh I see. Let's do `cast_list_f16`?

I want to limit how useful this function is as it's opaque to the optimizer. if someone calls `cast_f16(f16)`, it's be a bunch of work to make the optimizer understand that it is no-op

Comment (eddyxu) [reply rust/lance/src/dataset/scanner.rs:2895]: if we meant to have this as temporary solution, can we make it clear so? idk whether we can make a `_cast_list_f16` with leading underscore.

**Review by eddyxu — APPROVED** [review]

**Pull request merged**